### ISSUE TITLE
runserver + HTTPS doesn't work

### ISSUE THREAD

**Issue by kylemacfarlane** [jrief/django-websocket-redis#132]
I'm using the current master with Python 3 support.

I have HSTS enabled on our domain so even our development subdomains have to use HTTPS. This is very easy to achieve by wrapping Django's runserver.

However HTTPS doesn't play well with django-websocket-redis during development (I would presume it's fine behind nginx). WebSockets can connect but as soon as data is received they disconnect due to bad headers. This seems to be because the WebSockets passed around by the WSGI handler are just instances of io.BufferedReader and have no concept of HTTPS.

It's very easy to get a HTTPS WebSocket server running with pywsgi from gevent and gevent-websocket (+ some Python 3 fixes) but then I still can't seem to get the django-websocket-redis main loop to work properly. The problem seems to be that the way django-redis-websocket juggles file descriptors doesn't work with how pywsgi juggles sockets. There also doesn't seem to be any way to detect disconnects.

Below is my current code:

    def _runserver():
        from django.core.handlers.wsgi import WSGIHandler as DjangoListener
        from django.conf import settings
        from gevent import select, signal as g_signal
        from gevent.pool import Pool
        from gevent.pywsgi import WSGIServer
        from geventwebsocket.handler import WebSocketHandler
        from geventwebsocket.exceptions import WebSocketError
        import signal
        from ws4redis.exceptions import WebSocketError as RedisError
        from ws4redis.wsgi_server import WebsocketWSGIServer as RedisListener

        class WebSocket:
            def __init__(self, ws):
                self._closed = False
                self.ws = ws

            def get_file_descriptor(self):
                return self.ws.stream.handler.socket.fileno()

            @property
            def closed(self):
                return self._closed

            def receive(self):
                if self._closed:
                    raise RedisError('Connection is already closed')
                try:
                    data = self.ws.receive()
                    if data is not None:
                        return data.encode('utf-8')
                except WebSocketError as e:
                    self.close()
                    raise RedisError(e)

            def flush(self):
                pass

            def send(self, message, binary=None):
                self.ws.send(message, binary)

            def close(self, code=1000, message=''):
                self.ws.close(code, message)
                self._closed = True

        class GeventListener(RedisListener):
            def upgrade_websocket(self, environ, start_response):
                return WebSocket(environ['wsgi.websocket'])

            def select(self, rlist, wlist, xlist, timeout=None):
                return select.select(rlist, wlist, xlist, timeout)

        g_listener = GeventListener()
        _websocket_url = getattr(settings, 'WEBSOCKET_URL')

        class Listener(DjangoListener):
            def __call__(self, environ, start_response):
                if _websocket_url and environ.get('PATH_INFO').startswith(_websocket_url):
                    return g_listener(environ, start_response)
                return super().__call__(environ, start_response)

        bind = ('0.0.0.0', 8080)
        server = WSGIServer(
            bind,
            Listener(),
            spawn=Pool(10),
            handler_class=WebSocketHandler,
            certfile='cert.pem',
            keyfile='key.pem'
        )

        def stop(signal):
            if server and server.started:
                server.stop()
            sys.exit(signal)

        g_signal(signal.SIGINT, stop, signal.SIGINT)
        g_signal(signal.SIGTERM, stop, signal.SIGTERM)

        print('Starting WebSocket Enabled Server on %s:%s' % bind)
        server.serve_forever()

**Comment by jrief** [comment]
> However HTTPS doesn't play well with django-websocket-redis during development

Does this mean, that your ``./manage.py runserver`` speak HTTPS to your browser? Never even thought about that possibility.

Why don't you use pure uWSGI? This daemon is able to speak HTTPS. 
If you use this configuration: http://django-websocket-redis.readthedocs.org/en/latest/running.html#django-with-websockets-for-redis-as-a-stand-alone-uwsgi-server-in-emperor-mode
and add the private key/certificate to uWSGI, in my opinion it should work. Never tried tough. Let me know, if that works for you.

**Comment by kylemacfarlane** [comment]
Yeah that's exactly what my run command does.

I use uWSGI in production but for developement it's much easier to have everything in pure Python. Especially inside a virtual machine where inotify events don't work properly and thus neither does uWSGI file change detection.

The problem seems to be that the WSGI handler in Django wasn't designed to do either HTTPS or WebSockets and while it can do both individually it can't do both at the same time. Thus using the WSGI stuff in gevent is the easiest way to get going but that doesn't work django-websocket-redis.

If django-websocket-redis was less tied to uWSGI and runserver it would be easy enough to just plug gevent-websocket in. The big problem is the select method which appears to be tightly tied to traditional IO? If the select method had access to the environ then you could just pass around the WebSocket itself.

**Comment by kylemacfarlane** [comment]
I don't need this feature anymore but other might. In fact I didn't end up using django-websocket-redis at all.

Once I saw how easy it was to use websockets with uWSGI I moved over my runserver command to uWSGI as well (I don't use the built in runserver for dev due to previously mentioned SSL + inotify issues).

[This recent patch to uWSGI](https://github.com/unbit/uwsgi/pull/986) makes it super easy use uWSGI websockets with gevent for a completely non blocking main loop.

To handle auth I use an XMLHttpRequest before opening the websocket to obtain a short lived token from a Django view that is then stored in redis and given to the user (I got this idea from playing with services like PubNub). Thus my main loop never hits in the database - in fact it's only imports are gevent, uwsgi, redis, and the stdlib.

Also inspired by [this SO question](http://stackoverflow.com/questions/7196306/competing-consumer-on-redis-pub-sub-supported), instead of just using publish and listen I use rpush/publish and listen/lpop so that I can have multiple consumers without collisions. Thus redis handles all the atomic locking stuff so again it's all super simple.
